### PR TITLE
RavenDB-14614: add retention time for IoMetrics

### PR DESCRIPTION
**Pull request by garayx**

### CODE REVIEW

File: src/Raven.Server/Config/Categories/DatabaseConfiguration.cs [hunk 1]
```diff
@@ -76,5 +76,13 @@ public class DatabaseConfiguration : ConfigurationCategory
...
         [ConfigurationEntry("Databases.PulseReadTransactionLimitInMb", ConfigurationEntryScope.ServerWideOrPerDatabase)]
         public Size PulseReadTransactionLimit { get; set; }
+
+        /// <summary>
+        /// Specifies the time interval between each IoMetrics Cleaner run
+        /// </summary>
+        [Description("Retention time (in hours) for IO Metrics data")]
+        [DefaultValue(4)]
+        [ConfigurationEntry("Databases.IoMetricsCleanTimeInterval", ConfigurationEntryScope.ServerWideOrPerDatabase)]
+        public int IoMetricsCleanTimeInterval { get; set; }
```
Comment (grisha-kotler) [inline src/Raven.Server/Config/Categories/DatabaseConfiguration.cs]:
```suggestion
        public TimeSetting IoMetricsCleanTimeInterval { get; set; }
```

File: src/Raven.Server/Documents/DocumentDatabase.cs [hunk 2]
```diff
@@ -171,6 +174,7 @@ public DocumentDatabase(string name, RavenConfiguration configuration, ServerSto
         public ServerStore ServerStore => _serverStore;
 
         public DateTime LastIdleTime => new DateTime(_lastIdleTicks);
+        public DateTime IoMetricsCleanTime => new DateTime(_ioMetricsCleanerTicks);
```
Comment (grisha-kotler) [inline src/Raven.Server/Documents/DocumentDatabase.cs]: This isn't being used?

File: src/Raven.Server/Documents/DocumentDatabase.cs [hunk 3]
```diff
@@ -99,6 +101,7 @@ public DocumentDatabase(string name, RavenConfiguration configuration, ServerSto
             _logger = LoggingSource.Instance.GetLogger<DocumentDatabase>(Name);
             _serverStore = serverStore;
             _addToInitLog = addToInitLog;
+            _ioMetricsCleanerTicks = DateTime.UtcNow.AddHours(Configuration.Databases.IoMetricsCleanTimeInterval).Ticks;
```
Comment (grisha-kotler) [inline src/Raven.Server/Documents/DocumentDatabase.cs]: Save the `_lastMetriceCleanerRun` in `DateTime` format

**Review by grisha-kotler — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Documents/DocumentDatabase.cs (hunk 2 above)
Comment (garayx) [reply src/Raven.Server/Documents/DocumentDatabase.cs]: no, will delete it

File: src/Raven.Server/Config/Categories/DatabaseConfiguration.cs [hunk 4]
```diff
@@ -76,5 +76,13 @@ public class DatabaseConfiguration : ConfigurationCategory
         [SizeUnit(SizeUnit.Megabytes)]
         [ConfigurationEntry("Databases.PulseReadTransactionLimitInMb", ConfigurationEntryScope.ServerWideOrPerDatabase)]
         public Size PulseReadTransactionLimit { get; set; }
+
+        /// <summary>
+        /// Specifies the time interval between each IoMetrics Cleaner run
+        /// </summary>
+        [Description("Retention time (in hours) for IO Metrics data")]
+        [DefaultValue(4)]
```
Comment (grisha-kotler) [inline src/Raven.Server/Config/Categories/DatabaseConfiguration.cs]: Missing `[TimeUnit(TimeUnit.Hours)]`
Let's make it 24 hours by default

**Review by grisha-kotler — APPROVED** [review]

File: src/Raven.Server/Config/Categories/DatabaseConfiguration.cs [hunk 5]
```diff
@@ -76,5 +76,14 @@ public class DatabaseConfiguration : ConfigurationCategory
...
         public Size PulseReadTransactionLimit { get; set; }
+
+        /// <summary>
+        /// Specifies the time interval between each IoMetrics Cleaner run
+        /// </summary>
+        [Description("Retention time (in hours) for IO Metrics data")]
+        [DefaultValue(24)]
+        [TimeUnit(TimeUnit.Hours)]
+        [ConfigurationEntry("Databases.IoMetricsCleanTimeInterval", ConfigurationEntryScope.ServerWideOrPerDatabase)]
+        public TimeSetting IoMetricsCleanTimeInterval { get; set; }
```
Comment (ppekrol) [inline src/Raven.Server/Config/Categories/DatabaseConfiguration.cs]: Should go to `StorageConfiguration` IMO
and key should be `Storage.IoMetricsCleanupIntervalInHrs`

File: src/Raven.Server/Documents/DocumentDatabase.cs [hunk 6]
```diff
@@ -80,6 +81,7 @@ public class DocumentDatabase : IDisposable
         private long _usages;
         private readonly ManualResetEventSlim _waitForUsagesOnDisposal = new ManualResetEventSlim(false);
         private long _lastIdleTicks = DateTime.UtcNow.Ticks;
+        private DateTime _ioMetricsCleanerRun;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/DocumentDatabase.cs]: Rename to: `_nextIoMetricsCleanupTime`

File: src/Raven.Server/Documents/DocumentDatabase.cs [hunk 7]
```diff
@@ -906,17 +909,29 @@ public void RunIdleOperations()
...
+                {
+                    IoMetricsUtil.CleanIoMetrics(GetAllStoragesEnvironment(), _ioMetricsCleanerRun.Ticks);
+                    _ioMetricsCleanerRun = utcNow.Add(Configuration.Databases.IoMetricsCleanTimeInterval.AsTimeSpan);
+                    ranIoMetricsCleaner = true;
+                }
+
                 if (_logger.IsInfoEnabled)
-                    _logger.Info($"Ran idle operations for database '{Name}', took: {sp.ElapsedMilliseconds}ms");
+                {
+                    var msg = ranIoMetricsCleaner ? " with IO Metrics Cleaner" : string.Empty;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/DocumentDatabase.cs]: Why is this so important?

File: src/Raven.Server/Config/Categories/DatabaseConfiguration.cs [hunk 8]
```diff
@@ -76,5 +76,14 @@ public class DatabaseConfiguration : ConfigurationCategory
         [SizeUnit(SizeUnit.Megabytes)]
         [ConfigurationEntry("Databases.PulseReadTransactionLimitInMb", ConfigurationEntryScope.ServerWideOrPerDatabase)]
         public Size PulseReadTransactionLimit { get; set; }
+
+        /// <summary>
+        /// Specifies the time interval between each IoMetrics Cleaner run
+        /// </summary>
+        [Description("Retention time (in hours) for IO Metrics data")]
```
Comment (ppekrol) [inline src/Raven.Server/Config/Categories/DatabaseConfiguration.cs]: `Time (in hours) between IO Metrics cleanup`

File: src/Raven.Server/Utils/IoMetrics/IoMetricsUtil.cs [hunk 9]
```diff
@@ -107,8 +107,21 @@ public static IOMetricsRecentStats GetIoMetricsRecentStats(IoMeterBuffer.MeterIt
                 Type = recentMetric.Type
             };
         }
-    }
 
+        public static void CleanIoMetrics(IEnumerable<StorageEnvironmentWithType> environments, long ioMetricsCleanerTicks)
```
Comment (ppekrol) [inline src/Raven.Server/Utils/IoMetrics/IoMetricsUtil.cs]: Rename `ioMetricsCleanerTicks` to `untilTicks`

File: src/Raven.Server/Documents/DocumentDatabase.cs (hunk 7 above)
Comment (garayx) [reply src/Raven.Server/Documents/DocumentDatabase.cs]: removed

File: src/Raven.Server/Config/Categories/StorageConfiguration.cs [hunk 10]
```diff
@@ -85,6 +85,15 @@ public class StorageConfiguration : ConfigurationCategory
...
         public int SyncJournalsCountThreshold { get; set; }
 
+        /// <summary>
+        /// Specifies the time interval between each IoMetrics Cleaner run
+        /// </summary>
+        [Description("Time (in hours) between IO Metrics cleanup")]
+        [DefaultValue(24)]
+        [TimeUnit(TimeUnit.Hours)]
+        [ConfigurationEntry("Storage.IoMetricsCleanupIntervalInHrs", ConfigurationEntryScope.ServerWideOrPerDatabase)]
+        public TimeSetting IoMetricsCleanupIntervalInHrs { get; set; }
```
Comment (ppekrol) [inline src/Raven.Server/Config/Categories/StorageConfiguration.cs]: property does not have to have 'InHrs' in contract to configuration key. It is a TimeSetting with TimeSpan value

Comment (ppekrol) [reply src/Raven.Server/Config/Categories/StorageConfiguration.cs]: Done

**Review by ppekrol — APPROVED** [review]

**Pull request merged**